### PR TITLE
Batch progress

### PR DESCRIPTION
**Pull request by irevoire**
# Pull Request

## Related issue
Fixes https://github.com/meilisearch/meilisearch/issues/5068

## What does this PR do?
- ...

## PR checklist
Please check if your PR fulfills the following requirements:
- [ ] Does this PR fix an existing issue, or have you listed the changes applied in the PR description (and why they are needed)?
- [ ] Have you read the contributing guidelines?
- [ ] Have you made sure that the title is accurate and descriptive of the changes?

Thank you so much for contributing to Meilisearch!

### CODE REVIEW

File: crates/milli/src/progress.rs [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+pub trait NamedStep: 'static + Send + Sync + Default {
+    fn name(&self) -> &'static str;
+}
+
+/// Structure to quickly define steps that need very quick, lockless updating of their current step.
+/// You can use this struct if:
+/// - The name of the step doesn't change
+/// - The total number of steps doesn't change
+pub struct AtomicSubStep<Name: NamedStep> {
+    name: Name,
```
Comment (dureuill) [inline crates/milli/src/progress.rs:68]: consider renaming to `unit_name`?

File: crates/index-scheduler/src/batch.rs [hunk 2]
```diff
@@ -675,18 +690,29 @@ impl IndexScheduler {
                 fs::create_dir_all(&update_files_dir)?;
 
                 // 2.4 Only copy the update files of the enqueued tasks
+                progress.update_progress(SnapshotCreationProgress::SnapshotTheUpdateFiles);
+                let enqueued = self.get_status(&rtxn, Status::Enqueued)?;
+                let (atomic, update_file_progress) =
+                    AtomicUpdateFileStep::new(enqueued.len() as u32);
+                progress.update_progress(update_file_progress);
                 for task_id in self.get_status(&rtxn, Status::Enqueued)? {
```
Comment (dureuill) [inline crates/index-scheduler/src/batch.rs:698]: Reuse `enqueued` here

File: crates/milli/src/update/new/indexer/document_operation.rs [hunk 3]
```diff
@@ -68,16 +71,14 @@ impl<'pl> DocumentOperation<'pl> {
         let mut primary_key = None;
 
         let payload_count = operations.len();
+        let (step, progress_step) = AtomicDocumentStep::new(payload_count as u32);
```
Comment (dureuill) [inline crates/milli/src/update/new/indexer/document_operation.rs:74]: These are payloads, not documents

File: crates/index-scheduler/src/batch.rs [hunk 4]
```diff
@@ -1300,6 +1359,7 @@ impl IndexScheduler {
                     }
                 };
 
+                progress.update_progress(DocumentOperationProgress::ComputingTheChanges);
```
Comment (dureuill) [inline crates/index-scheduler/src/batch.rs:1362]:
```suggestion
                progress.update_progress(DocumentOperationProgress::ComputingDocumentChanges);
```

File: crates/index-scheduler/src/batch.rs [hunk 5]
```diff
@@ -1446,6 +1509,7 @@ impl IndexScheduler {
                     };
 
                     let candidates_count = candidates.len();
+                    progress.update_progress(DocumentEditionProgress::ComputingTheChanges);
```
Comment (dureuill) [inline crates/index-scheduler/src/batch.rs:1512]:
```suggestion
                    progress.update_progress(DocumentEditionProgress::ComputingDocumentChanges);
```

File: crates/index-scheduler/src/processing.rs [hunk 6]
```diff
@@ -0,0 +1,356 @@
...
+        RetrievingConfig,
+        ComputingTheChanges,
+        Indexing,
+    }
+}
+
+make_enum_progress! {
+    pub enum DocumentEditionProgress {
+        RetrievingConfig,
+        ComputingTheChanges,
```
Comment (dureuill) [inline crates/index-scheduler/src/processing.rs:191]:
```suggestion
        ComputingDocumentChanges,
```

File: crates/index-scheduler/src/batch.rs [hunk 7]
```diff
@@ -1805,6 +1899,7 @@ impl IndexScheduler {
                     }
                 }
             }
+            atomic_progress.fetch_add(1, Ordering::Relaxed);
```
Comment (dureuill) [inline crates/index-scheduler/src/batch.rs:1902]: consider inspecting the iterator to perform the fetch add while iterating (more robust in the face of early continues)

File: crates/index-scheduler/src/utils.rs [hunk 8]
```diff
@@ -273,7 +275,10 @@ impl IndexScheduler {
             .into_iter()
             .map(|batch_id| {
                 if Some(batch_id) == processing.batch.as_ref().map(|batch| batch.uid) {
-                    Ok(processing.batch.as_ref().unwrap().to_batch())
+                    let mut batch = processing.batch.as_ref().unwrap().to_batch();
+                    println!("here with progress: {}", processing.progress.is_some());
```
Comment (dureuill) [inline crates/index-scheduler/src/utils.rs:279]:
```suggestion
```

File: crates/milli/src/update/new/indexer/document_changes.rs [hunk 9]
```diff
@@ -213,28 +206,30 @@ pub fn extract<
         doc_allocs,
         fields_ids_map_store,
         must_stop_processing,
-        send_progress,
-    }: IndexingContext<'fid, 'indexer, 'index, MSP, SP>,
+        progress: send_progress,
```
Comment (dureuill) [inline crates/milli/src/update/new/indexer/document_changes.rs:209]:
```suggestion
        progress,
```

File: crates/milli/src/update/new/indexer/document_changes.rs [hunk 10]
```diff
@@ -213,28 +206,30 @@ pub fn extract<
...
-    step: Step,
+    step: IndexingStep,
 ) -> Result<()>
 where
     EX: Extractor<'extractor>,
     MSP: Fn() -> bool + Sync,
-    SP: Fn(Progress) + Sync,
 {
     tracing::trace!("We are resetting the extractor allocators");
+    send_progress.update_progress(step);
```
Comment (dureuill) [inline crates/milli/src/update/new/indexer/document_changes.rs:220]:
```suggestion
    progress.update_progress(step);
```

File: crates/milli/src/update/new/indexer/document_changes.rs [hunk 11]
```diff
@@ -213,28 +206,30 @@ pub fn extract<
...
     // Clean up and reuse the extractor allocs
     for extractor_alloc in extractor_allocs.iter_mut() {
         tracing::trace!("\tWith {} bytes reset", extractor_alloc.0.allocated_bytes());
         extractor_alloc.0.reset();
     }
 
     let total_documents = document_changes.len() as u32;
+    let (step, progress_step) = AtomicDocumentStep::new(total_documents);
+    send_progress.update_progress(progress_step);
```
Comment (dureuill) [inline crates/milli/src/update/new/indexer/document_changes.rs:229]:
```suggestion
    progress.update_progress(progress_step);
```

File: crates/milli/src/progress.rs [hunk 12]
```diff
@@ -0,0 +1,116 @@
...
+pub struct ProgressView {
+    steps: Vec<ProgressStepView>,
+    percentage: f32,
+}
+
+#[derive(Debug, Serialize, Clone)]
+pub struct ProgressStepView {
+    name: Cow<'static, str>,
+    finished: u32,
+    total: u32,
```
Comment (dureuill) [inline crates/milli/src/progress.rs:115]:
```suggestion
    pub name: Cow<'static, str>,
    pub finished: u32,
    pub total: u32,
```

File: crates/milli/src/progress.rs [hunk 13]
```diff
@@ -0,0 +1,116 @@
...
+        "document"
+    }
+}
+
+pub type AtomicDocumentStep = AtomicSubStep<Document>;
+
+#[derive(Debug, Serialize, Clone)]
+pub struct ProgressView {
+    steps: Vec<ProgressStepView>,
+    percentage: f32,
```
Comment (dureuill) [inline crates/milli/src/progress.rs:108]:
```suggestion
    pub steps: Vec<ProgressStepView>,
    pub percentage: f32,
```

File: crates/milli/src/progress.rs [hunk 14]
```diff
@@ -0,0 +1,116 @@
...
+    }
+
+    // TODO: This code should be in meilisearch_types but cannot because milli can't depend on meilisearch_types
+    pub fn as_progress_view(&self) -> ProgressView {
+        let steps = self.steps.read().unwrap();
+
+        let mut percentage = 0.0;
+        let mut prev_factors = 1.0;
+
+        let mut step_view = Vec::new();
```
Comment (dureuill) [inline crates/milli/src/progress.rs:40]:
```suggestion
        let mut step_view = Vec::with_capacity(steps.len());
```

File: crates/milli/src/progress.rs [hunk 15]
```diff
@@ -0,0 +1,116 @@
...
+pub type AtomicDocumentStep = AtomicSubStep<Document>;
+
+#[derive(Debug, Serialize, Clone)]
+pub struct ProgressView {
+    steps: Vec<ProgressStepView>,
+    percentage: f32,
+}
+
+#[derive(Debug, Serialize, Clone)]
+pub struct ProgressStepView {
```
Comment (dureuill) [inline crates/milli/src/progress.rs:112]:
```suggestion
#[serde(rename_all="camelCase")]
pub struct ProgressStepView {
```

File: crates/meilisearch-types/src/batches.rs [hunk 16]
```diff
@@ -15,6 +16,8 @@ pub type BatchId = u32;
 pub struct Batch {
     pub uid: BatchId,
 
+    #[serde(skip_deserializing)]
```
Comment (dureuill) [inline crates/meilisearch-types/src/batches.rs:19]:
```suggestion
    #[serde(skip)]
```

**Review by dureuill — COMMENTED** [review]
Some minor comments, approval should come quickly after handling them 

Thank you!🌈

File: crates/index-scheduler/src/batch.rs (hunk 7 above)
Comment (irevoire) [reply crates/index-scheduler/src/batch.rs:1902]: In the end we said it wasn't a good idea since

**Review by dureuill — APPROVED** [review]
game changer feature ❤️‍🔥 

bors merge

**Pull request merged**